subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CopyTaskIntegrationSpec.groovy

```diff
@@ -2239,4 +2242,121 @@ class CopyTaskIntegrationSpec extends AbstractIntegrationSpec {
+        TestFile zipFile = file("before/files.zip")
+        unzippedDir.zipTo(zipFile)
+
+        and: "another zip file with the same contents"
+        TestFile zipFile2 = file("before/files2.zip")
+        unzippedDir.zipTo(zipFile2)
+
+        and: "a copy task that copies from both zip files, failing on duplicates"
+        buildFile << """
+            task (copy, type: Copy) {
+                from zipTree('${separatorsToUnix(zipFile.path)}')
+                from zipTree('${separatorsToUnix(zipFile2.path)}')
+                into 'after'
+                duplicatesStrategy = DuplicatesStrategy.FAIL
+            }
+        """
+
+        expect: "a failure"
+        fails 'copy'
+
+        and: "with a clear failure message"
+        failure.assertHasCause('Encountered duplicate path "sub/c.txt", extracted from "files2.zip", during copy operation configured with DuplicatesStrategy.FAIL')
+        failure.assertHasDescription("Execution failed for task ':copy'.")
+    }
+
+    def "encountering duplicates in a tarTree vs an uncompressed dir with DuplicateStrategy.FAIL should give a clear error"() {
+        given: "a directory with a file"
+        TestFile untarredDir = file("before/files")
+        untarredDir.file("sub/c.txt").createFile()
+
+        and: "a tar file containing it"
+        TestFile tarFile = file("before/files.tar")
+        untarredDir.tarTo(tarFile)
+
+        and: "a copy task that copies from both of these, failing on duplicates"
+        buildFile << """
+            task (copy, type: Copy) {
+                from '${separatorsToUnix(untarredDir.path)}'
+                from tarTree('${separatorsToUnix(tarFile.path)}')
+                into 'after'
+                duplicatesStrategy = DuplicatesStrategy.FAIL
+            }
+        """
+
+        expect: "a failure"
+        fails 'copy'
+
+        and: "with a clear failure message"
+        failure.assertHasCause('Encountered duplicate path "sub/c.txt", extracted from "files.tar", during copy operation configured with DuplicatesStrategy.FAIL')
+        failure.assertHasDescription("Execution failed for task ':copy'.")
+    }
+
+    def "encountering duplicates in a tarTree vs another tarTree with DuplicateStrategy.FAIL should give a clear error"() {
+        given: "a directory with a file"
+        TestFile untarredDir = file("before/files")
+        untarredDir.file("sub/c.txt").createFile()
+
+        and: "a tar file containing it"
+        TestFile tarFile = file("before/files.tar")
+        untarredDir.tarTo(tarFile)
+
+        and: "another tar file with the same contents"
+        TestFile tarFile2 = file("before/files2.tar")
+        untarredDir.tarTo(tarFile2)
+
+        and: "a copy task that copies from both tar files, failing on duplicates"
+        buildFile << """
+            task (copy, type: Copy) {
+                from tarTree('${separatorsToUnix(tarFile.path)}')
+                from tarTree('${separatorsToUnix(tarFile2.path)}')
```

Comment:
🤔 is there any reason to believe that the combination of two tarTrees could behave differently from a tarTree + directory?  Why not a zipTree + tarTree? I think we might be able to get rid of this test.

Reply:
I wasn't sure if the Copy task would somehow optimize thinking "this is the same source, I can skip warning about this this" OR if it would de-duplicate source paths prior to copying.  I can delete these tests if you'd like.